### PR TITLE
Add TidbMonitor Controller

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 

resolve: https://github.com/pingcap/tidb-operator/issues/1261 controller part.

### CODE REVIEW

File: pkg/monitor/monitor/util.go [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+			Namespace:       monitor.Namespace,
+			Labels:          monitorLabel,
+			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
+			Annotations:     monitor.Spec.Annotations,
+		},
+		Spec: apps.DeploymentSpec{
+			Replicas: &replicas,
+			Strategy: apps.DeploymentStrategy{
+				Type:          apps.RecreateDeploymentStrategyType,
+				RollingUpdate: nil,
```
Comment (aylei) [inline pkg/monitor/monitor/util.go]: I think the BP is to omit unnecessary fields: https://kubernetes.io/docs/concepts/configuration/overview/#general-configuration-tips

File: pkg/monitor/monitor/util.go [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+		},
+	}
+	return deployment
+}
+
+func getMonitorInitContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.TidbCluster) core.Container {
+	c := "mkdir -p /data/prometheus\nchmod 777 /data/prometheus\n/usr/bin/init.sh"
+	if monitor.Spec.Grafana != nil {
+		c = "mkdir -p /data/prometheus /data/grafana\nchmod 777 /data/prometheus /data/grafana\n/usr/bin/init.sh"
+	}
```
Comment (aylei) [inline pkg/monitor/monitor/util.go]: Write multiple-line string in raw string literals is more readable

File: pkg/monitor/monitor/util.go [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+				Value: tc.Namespace,
+			},
+			{
+				Name:  "TZ",
+				Value: tc.Spec.Timezone,
+			},
+		},
+		Command: command,
+		SecurityContext: &core.SecurityContext{
+			RunAsUser: &secureContext,
```
Comment (aylei) [inline pkg/monitor/monitor/util.go]: we can use `pointer.Int64(0)`

File: pkg/monitor/monitor/util.go [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+		})
+	}
+	return c
+}
+
+func getMonitorReloaderContainer(monitor *v1alpha1.TidbMonitor, tc *v1alpha1.TidbCluster) core.Container {
+	c := core.Container{
+		Name:            "reloader",
+		Image:           fmt.Sprintf("%s:%s", monitor.Spec.Reloader.BaseImage, monitor.Spec.Reloader.Version),
+		ImagePullPolicy: *monitor.Spec.Reloader.ImagePullPolicy,
```
Comment (aylei) [inline pkg/monitor/monitor/util.go]: check nil

File: pkg/monitor/monitor/util.go [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+	}
+	volumes = append(volumes, prometheusRules)
+	if tc.Spec.EnableTLSCluster {
+		defaultMode := int32(420)
+		tlsPDClient := core.Volume{
+			Name: "tls-pd-client",
+			VolumeSource: core.VolumeSource{
+				Secret: &core.SecretVolumeSource{
+					SecretName:  fmt.Sprintf("%s-pd-client", tc.Name),
+					DefaultMode: &defaultMode,
```
Comment (aylei) [inline pkg/monitor/monitor/util.go]: ditto

File: cmd/controller-manager/main.go [hunk 6]
```diff
@@ -16,6 +16,7 @@ package main
 import (
 	"context"
 	"flag"
+	"github.com/pingcap/tidb-operator/pkg/controller/tidbmonitor"
```
Comment (shuijing198799) [inline cmd/controller-manager/main.go]: move this package to the second part with prefix github.com

Comment (Yisaer) [reply cmd/controller-manager/main.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+// limitations under the License.
+
+package monitor
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	informers "github.com/pingcap/tidb-operator/pkg/client/informers/externalversions"
+	v1alpha1listers "github.com/pingcap/tidb-operator/pkg/client/listers/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	core "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: pkg/controller/tidbmonitor/tidb_monitor_control.go [hunk 8]
```diff
@@ -27,22 +30,34 @@ type ControlInterface interface {
...
+	oldStatus := tm.Status.DeepCopy()
 
-	// TODO: implementation
+	if err := tmc.reconcileTidbMonitor(tm); err != nil {
+		errs = append(errs, err)
+	}
+	if apiequality.Semantic.DeepEqual(&tm.Status, oldStatus) {
+		return errorutils.NewAggregate(errs)
+	}
 	return errors.NewAggregate(errs)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/controller/tidbmonitor/tidb_monitor_control.go]: updated and removed.

File: pkg/controller/tidbmonitor/tidb_monitor_control.go [hunk 9]
```diff
@@ -16,7 +16,10 @@ package tidbmonitor
 import (
 	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
 	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/monitor"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/util/errors"
+	errorutils "k8s.io/apimachinery/pkg/util/errors"
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/controller/tidbmonitor/tidb_monitor_control.go]: updated.

File: pkg/monitor/monitor/util.go (hunk 1 above)
Comment (Yisaer) [reply pkg/monitor/monitor/util.go]: updated.

File: pkg/monitor/monitor/util.go (hunk 3 above)
Comment (Yisaer) [reply pkg/monitor/monitor/util.go]: updated.

File: pkg/monitor/monitor/util.go (hunk 4 above)
Comment (Yisaer) [reply pkg/monitor/monitor/util.go]: updated.

File: pkg/monitor/monitor/util.go (hunk 5 above)
Comment (Yisaer) [reply pkg/monitor/monitor/util.go]: updated.

File: pkg/monitor/monitor/util.go (hunk 2 above)
Comment (Yisaer) [reply pkg/monitor/monitor/util.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+	if monitor.DeletionTimestamp != nil {
+		return nil
+	}
+
+	// Sync Service
+	if err := mm.syncTidbMonitorService(monitor); err != nil {
+		return err
+	}
+	klog.Infof("tm[%s/%s]'s service synced", monitor.Namespace, monitor.Name)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (mm *MonitorManager) syncTidbMonitorDeployment(monitor *v1alpha1.TidbMonitor) error {
+	targetTcRef := monitor.Spec.Clusters[0]
+	tc, err := mm.tcLister.TidbClusters(targetTcRef.Namespace).Get(targetTcRef.Name)
+	if err != nil {
+		return err
+	}
+
+	cm, err := mm.syncTidbMonitorConfig(tc, monitor)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: pkg/controller/generic_control.go [hunk 12]
```diff
@@ -66,6 +73,71 @@ func NewTypedControl(control GenericControlInterface) TypedControlInterface {
...
 }
 
+func (w *typedWrapper) CreateOrUpdatePVC(controller runtime.Object, pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	result, err := w.GenericControlInterface.CreateOrUpdate(controller, pvc, func(existing, desired runtime.Object) error {
+		existingPVC := existing.(*corev1.PersistentVolumeClaim)
+		desiredPVC := desired.(*corev1.PersistentVolumeClaim)
+
+		existingPVC.Labels = desiredPVC.Labels
+		existingPVC.Spec = desiredPVC.Spec
+		existingPVC.Annotations = desiredPVC.Annotations
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/controller/generic_control.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	return mm.syncTidbMonitorDeployment(monitor)
+}
+
+func (mm *MonitorManager) syncTidbMonitorService(monitor *v1alpha1.TidbMonitor) error {
+	service := getMonitorService(monitor)
+	for _, svc := range service {
+		_, err := mm.typedControl.CreateOrUpdateService(monitor, svc)
+		if err != nil {
+			klog.Errorf("tm[%s/%s]'s service[%s] failed to sync,err: %v", monitor.Namespace, monitor.Name, svc.Name, err)
+			return err
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+	if exist {
+		return nil
+	}
+	pvc = getMonitorPVC(monitor)
+	err = mm.typedControl.Create(monitor, pvc)
+	if err != nil {
+		klog.Errorf("tm[%s/%s]'s pvc[%s] failed to sync,err: %v", monitor.Namespace, monitor.Name, pvc.Name, err)
+		return err
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: cmd/controller-manager/main.go [hunk 15]
```diff
@@ -21,6 +21,8 @@ import (
 	"os"
 	"time"
 
+	"github.com/pingcap/tidb-operator/pkg/controller/tidbmonitor"
+
```
Comment (shuijing198799) [inline cmd/controller-manager/main.go]: remove this line ……

Comment (Yisaer) [reply cmd/controller-manager/main.go]: This was formatted by `goimports`

File: pkg/monitor/monitor/monitor_manager.go [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+		return nil, err
+	}
+	crb := getMonitorClusterRoleBinding(sa, cr, monitor)
+	_, err = mm.typedControl.CreateOrUpdateClusterRoleBinding(monitor, crb)
+	if err != nil {
+		klog.Errorf("tm[%s/%s]'s clusterRile failed to sync,err: %v", monitor.Namespace, monitor.Name, err)
+		return nil, err
+	}
+	return sa, nil
+}
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go]: updated.

File: cmd/controller-manager/main.go (hunk 15 above)
Comment (Yisaer) [reply cmd/controller-manager/main.go]: updated.

File: pkg/monitor/monitor/template.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+  scrape_interval: 15s
+  evaluation_interval: 15s
+{{- if .AlertmanagerURL }}
+alerting:
+  alertmanagers:
+  - static_configs:
+    - targets:
+      - {{ .AlertmanagerURL }}
+{{- end }}
+scrape_configs:
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/template.go]: I tend to make the Go struct definition as a new issue.

[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/monitor/monitor/template.go]: Issued in https://github.com/pingcap/tidb-operator/issues/1434

File: pkg/monitor/monitor/monitor_manager.go [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+		cr := getMonitorClusterRole(monitor)
+		cr, err = mm.typedControl.CreateOrUpdateClusterRole(monitor, cr)
+		if err != nil {
+			klog.Errorf("tm[%s/%s]'s clusterrole failed to sync,err: %v", monitor.Namespace, monitor.Name, err)
+			return nil, err
+		}
+		crb := getMonitorClusterRoleBinding(sa, cr, monitor)
+		_, err = mm.typedControl.CreateOrUpdateClusterRoleBinding(monitor, crb)
+		if err != nil {
+			klog.Errorf("tm[%s/%s]'s clusterRile failed to sync,err: %v", monitor.Namespace, monitor.Name, err)
```
Comment (aylei) [inline pkg/monitor/monitor/monitor_manager.go]:
```suggestion
			klog.Errorf("tm[%s/%s]'s clusterRoleBinding failed to sync,err: %v", monitor.Namespace, monitor.Name, err)
```

**Pull request merged**